File: 지찬/프린터.java
//https://school.programmers.co.kr/learn/courses/30/lessons/42587
import java.util.Comparator;
import java.util.LinkedList;
import java.util.PriorityQueue;
import java.util.Queue;

class Solution {
    public int solution(int[] priorities, int location) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Comparator.reverseOrder());
        Queue<Integer> queue = new LinkedList<>();
        for (int p :
                priorities) {
            pq.add(p);
            queue.add(p);
        }
        int count = 1;
        while (!queue.isEmpty()) {
            if (pq.peek() == queue.peek()) {
                if (location == 0) {
                    return count;
                }
                pq.poll();
                queue.poll();
                count++;
                location--;
            } else {
                queue.add(queue.poll());
                if (location == 0) {
                    location = queue.size() - 1;
                } else {
                    location--;
                }
            }
        }
        return count;
    }

    public static void main(String[] args) {
        int[] priorities = {
                2, 1, 3, 2
        };
        int location = 2;
        Solution s = new Solution();
        System.out.println(s.solution(priorities, location));

    }
}

File: 종우/프린터.py
# 0105_1
# https://school.programmers.co.kr/learn/courses/30/lessons/42587

# 내 풀이
def solution(priorities, location):
    
    find = (priorities[location], location)
    x = list(zip(priorities, [i for i in range(len(priorities))]))
    tmp = []
    
    while len(x) != 0:
        if max(x)[0] == x[0][0]:
            tmp.append(x[0])
            x.pop(0)
        else:
            x.append(x[0])
            x.pop(0)     
            
    answer=tmp.index(find)+1
    return answer

# 다른사람 풀이
# The any() function returns True if any element of an iterable is True. If not, it returns False.

def solution(priorities, location):
    queue =  [(i,p) for i,p in enumerate(priorities)]
    answer = 0
    while True:
        cur = queue.pop(0)
        if any(cur[1] < q[1] for q in queue):
            queue.append(cur)
        else:
            answer += 1
            if cur[0] == location:
                return answer